session_test.go

```diff
@@ -2116,6 +2116,66 @@ func (s *testSessionSuite) TestSubstringIndexExpr(c *C) {
 	c.Assert(err, IsNil)
 }
 
+func (s *testSessionSuite) TestSpecifyIndexPrefixLength(c *C) {
+	defer testleak.AfterTest(c)()
+	store := newStore(c, s.dbName)
+	se := newSession(c, store, s.dbName)
+	mustExecSQL(c, se, "drop table if exists t;")
+	mustExecSQL(c, se, "create table t (c1 int, c2 blob, c3 varchar(64));")
+	_, err := exec(c, se, "create index idx_c1 on t (c2);")
+	// ERROR 1170 (42000): BLOB/TEXT column 'c2' used in key specification without a key length
+	c.Assert(err, NotNil)
+
+	// TODO:
+	// _, err := exec(c, se, "create index idx_c1 on t (c1(5))")
+	// ERROR 1089 (HY000): Incorrect prefix key;
+	// the used key part isn't a string, the used length is longer than the key part,
+	// or the storage engine doesn't support unique prefix keys
+
+	mustExecSQL(c, se, "create index idx_c1 on t (c1);")
+	mustExecSQL(c, se, "create index idx_c2 on t (c2(3));")
+	mustExecSQL(c, se, "create unique index idx_c3 on t (c3(5));")
+
+	mustExecSQL(c, se, "insert into t values (3, 'abc', 'def');")
+	sql := "select c2 from t where c2 = 'abc';"
+	mustExecMatch(c, se, sql, [][]interface{}{{[]byte("abc")}})
+
+	mustExecSQL(c, se, "insert into t values (4, 'abcd', 'xxx');")
+	mustExecSQL(c, se, "insert into t values (4, 'abcf', 'yyy');")
+	sql = "select c2 from t where c2 = 'abcf';"
+	mustExecMatch(c, se, sql, [][]interface{}{{[]byte("abcf")}})
+	sql = "select c2 from t where c2 = 'abcd';"
+	mustExecMatch(c, se, sql, [][]interface{}{{[]byte("abcd")}})
+
+	mustExecSQL(c, se, "insert into t values (4, 'ignore', 'abcdeXXX');")
+	_, err = exec(c, se, "insert into t values (5, 'ignore', 'abcdeYYY');")
+	// ERROR 1062 (23000): Duplicate entry 'abcde' for key 'idx_c3'
+	c.Assert(err, NotNil)
+	sql = "select c3 from t where c3 = 'abcde';"
+	mustExecMatch(c, se, sql, [][]interface{}{})
+
+	mustExecSQL(c, se, "delete from t where c3 = 'abcdeXXX';")
+	mustExecSQL(c, se, "delete from t where c2 = 'abc';")
+
+	mustExecMatch(c, se, "select c2 from t where c2 > 'abcd';", [][]interface{}{{[]byte("abcf")}})
+	mustExecMatch(c, se, "select c2 from t where c2 < 'abcf';", [][]interface{}{{[]byte("abcd")}})
+	mustExecMatch(c, se, "select c2 from t where c2 >= 'abcd';", [][]interface{}{{[]byte("abcd")}, {[]byte("abcf")}})
+	mustExecMatch(c, se, "select c2 from t where c2 <= 'abcf';", [][]interface{}{{[]byte("abcd")}, {[]byte("abcf")}})
+	mustExecMatch(c, se, "select c2 from t where c2 != 'abc';", [][]interface{}{{[]byte("abcd")}, {[]byte("abcf")}})
+	mustExecMatch(c, se, "select c2 from t where c2 != 'abcd';", [][]interface{}{{[]byte("abcf")}})
+
+	mustExecSQL(c, se, "drop table if exists t1;")
+	mustExecSQL(c, se, "create table t1 (a int, b char(255), key(a, b(20)));")
+	mustExecSQL(c, se, "insert into t1 values (0, '1');")
+	mustExecSQL(c, se, "update t1 set b = b + 1 where a = 0;")
+	mustExecMatch(c, se, "select b from t1 where a = 0;", [][]interface{}{{[]byte("2")}})
+
```

Comment:
Add test cases for multiple column index which has prefix index column at first.
like `create table t2 (a text, b text, index (a(3), b(3)))`